### DISCUSSION TITLE
Avoiding multiple iterations

### DISCUSSION THREAD

**Discussion by stenci** [wtertinek/Linq2Acad#46] [Q&A] [answered]
Question: is it fair to expect that an iterable passed to a method will be iterated only once? _(honest question from someone that is learning C# and trying to figure out how it works, but especially what's the right way to do things)_

If no, then you can ignore this.
If yes, then please keep reading.

---

I spent too much time trying to figure out why I was getting the `Forgot to call Dispose?` message in the Debug window. Even when I finally identified what was causing it, it was the following loop, I still couldn't figure out why:
```c#
var acadDb = AcadDatabaseCore.Active(DwgManager.TopTransaction);
var lines = jointLines
    .Select(line1 => line1.Line())
    .Select(line2 => new Line(line2.StartPoint, line2.StartPoint + (line2.EndPoint - line2.StartPoint) / 3));
acadDb.CurrentSpace.AddRange(lines);
```

The problem was the `EntityContainer.AddRange(IEnumerable<Entity> entities, [...])`, which was iterating three times: once in the `Require.ElementsNotNull`, once in the loop with `Require.NewlyCreated` and once in the `AddInternal(entities, [...])`. Every iteration would create a `new Line`, but only the last one would be added to the database.

I fixed by converting the three iterating checks into one loop that checks and processes each item. I also modified `AddInternal` to work with one entity rather than an array, and I introduced a new `Require.ElementNotNull`.

Here are the three modified functions:
```c#
public ObjectId Add(Entity entity, bool setDatabaseDefaults = false)
{
  Require.NotDisposed(database.IsDisposed, nameof(AcadDatabase));
  Require.TransactionNotDisposed(transaction.IsDisposed);
  Require.ParameterNotNull(entity, nameof(entity));
  Require.NewlyCreated(entity, nameof(entity));

  return AddInternal(entity, setDatabaseDefaults);
}

public IReadOnlyCollection<ObjectId> AddRange(IEnumerable<Entity> entities, bool setDatabaseDefaults = false)
{
  Require.NotDisposed(database.IsDisposed, nameof(AcadDatabase));
  Require.TransactionNotDisposed(transaction.IsDisposed);
  Require.ParameterNotNull(entities, nameof(entities));

  var ids = new List<ObjectId>();
  foreach (var entity in entities)
  {
    Require.ElementNotNull(entity, nameof(entities));
    Require.NewlyCreated(entity, nameof(entity));
    ids.Add(AddInternal(entity, setDatabaseDefaults));
  }

  return ids;
}

private ObjectId AddInternal(Entity item, bool setDatabaseDefaults)
{
  var btr = (BlockTableRecord)transaction.GetObject(ID, OpenMode.ForWrite);

  if (setDatabaseDefaults)
  {
    item.SetDatabaseDefaults();
  }

  var id = btr.AppendEntity(item);
  transaction.AddNewlyCreatedDBObject(item, true);

  return id;
}
```

The I did something similar in other methods that call `Require.ElementsNotNull`.

You can see the changes here: https://github.com/stenci/Linq2Acad/commit/9e325f4b4373ac35298bc0fd8df5915610db92e7

I think that multiple iteration should be avoided, but I might be wrong. Am I overthinking this?

---

I don't know whether that line of linq code that was iterated multiple times looks good to a good C# developer. It doesn't look very readable to me, but I'm learning linq and I'm trying to use it as much as I can. This is not production code, I was debugging a function, I created a little loop to add some short lines coincident to some existing lines (the thick white lines in the snapshot), to see if my algorithm was collecting the correct lines:
![image](https://github.com/wtertinek/Linq2Acad/assets/5955495/c79f3111-c6f1-4a1c-ad1c-6ea46e26a550)

**Comment by wtertinek** [comment, marked as answer]
Hi, this is the most annoying problem you encounter when you start with LINQ. LINQ queries are executed lazily, so a query is iterated every time you access it:

```csharp
IEnumerable<int> query = new [] { 1, 2, 3 }.Select(v => { Console.WriteLine(v); return v; });

Console.WriteLine("First iteration");

foreach (var value in query)
{
}

Console.WriteLine("Second iteration");

foreach (var value in query)
{
}
```

Here the result is:

First iteration
1
2
3
Second iteration
1
2
3

If you want to have it iterated only once, you have to **materialize** the query using .ToArray(), .ToList() or similar methods:

```csharp
int[] query = new [] { 1, 2, 3 }.Select(v => { Console.WriteLine(v); return v; }).ToArray();

Console.WriteLine("First iteration");

foreach (var value in query)
{
}

Console.WriteLine("Second iteration");

foreach (var value in query)
{
}
```

Here the result is:

1
2
3
First iteration
Second iteration

So in the second example the query is iterated only once by calling .ToArray() and converting it into an array.

It's common practice in C# to materialize your query as soon as possible. In your example I would simplay add a .ToArray() here:

```csharp
var lines = jointLines
    .Select(line1 => line1.Line())
    .Select(line2 => new Line(line2.StartPoint, line2.StartPoint + (line2.EndPoint - line2.StartPoint) / 3))
    .ToArray();
```

Hope that helps.
